File: a1rust/src/cmd.rs
use std::str::{FromStr, Split};

#[derive(Debug, PartialEq, Eq, Clone, Copy)]
pub enum Cmd {
    Alloc(usize, usize),
    Dealloc(usize),
    Compact,
    Output,
}

impl std::fmt::Display for Cmd {
    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
        match self {
            Cmd::Alloc(_, _) => write!(f, "A"),
            Cmd::Dealloc(_) => write!(f, "D"),
            Cmd::Compact => write!(f, "C"),
            Cmd::Output => write!(f, "O"),
        }
    }
}

impl FromStr for Cmd {
    type Err = std::io::Error;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        let iter = s.split(';');
        let (cmd, iter) = next_string(iter);
        match cmd.as_str() {
            "A" => {
                let (size, iter) = next_usize(iter);
                let (align, _) = next_usize(iter);
                Ok(Cmd::Alloc(size, align))
            }
            "D" => {
                let (id, _) = next_usize(iter);
                Ok(Cmd::Dealloc(id))
            }
            "O" => Ok(Cmd::Output),
            _ => Ok(Cmd::Compact),
        }
    }
}

fn next_string(mut iter: Split<char>) -> (String, Split<char>) {
    (iter.next().unwrap().to_owned(), iter)
}

fn next_usize(mut iter: Split<char>) -> (usize, Split<char>) {
    (iter.next().unwrap().parse().unwrap(), iter)
}

#[derive(Debug)]
pub struct CmdVec {
    pub size: usize,
    pub cmds: Vec<Cmd>,
}

impl CmdVec {
    pub fn new(size: usize) -> CmdVec {
        CmdVec { size, cmds: vec![] }
    }

    pub fn add(&mut self, cmd: Cmd) {
        self.cmds.push(cmd);
    }

    pub fn iter(&self) -> std::slice::Iter<Cmd> {
        self.cmds.iter()
    }
}

impl FromStr for CmdVec {
    type Err = std::io::Error;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        let lines = s.lines().collect::<Vec<&str>>();
        let size = lines.first().unwrap().parse().unwrap();
        let mut cmds = CmdVec::new(size);
        for line in lines[1..].iter() {
            let cmd = line.parse::<Cmd>()?;
            cmds.add(cmd);
        }
        Ok(cmds)
    }
}

File: a1/a1py/src/core/cmd.py
from enum import Enum
from typing import Optional
from dataclasses import dataclass


class Op(Enum):
    Alloc = "A"
    Dealloc = "D"
    Compact = "C"
    Output = "O"


@dataclass
class Cmd:
    op: Op
    id: Optional[int]
    size: Optional[int]

    def __init__(self, op: Op, id: Optional[int] = None, size: Optional[int] = None) -> None:
        self.op = op
        self.id = id
        self.size = size

    @staticmethod
    def from_str(line: str) -> "Cmd":
        parts: list[str] = [line[0]]
        if ";" in line:
            parts = line.split(";")
        match parts[0]:
            case "A":
                return Cmd(Op.Alloc, int(parts[1]), int(parts[2]))
            case "D":
                return Cmd(Op.Dealloc, int(parts[1]))
            case "C":
                return Cmd(Op.Compact)
            case "O":
                return Cmd(Op.Output)
            case _:
                raise ValueError("Invalid command")

    def get_id(self) -> int:
        if self.id is None:
            raise ValueError("Invalid command")
        return self.id

    def get_size(self) -> int:
        if self.size is None:
            raise ValueError("Invalid command")
        return self.size
